### PR TITLE
Better account name handling

### LINKED ISSUE #45: Invalid usernames still cause extension to hang
Looks like the current regex sometimes removes unrelated parts of the user object, which then makes the JSON invalid.

Current; 
image

### PR DESCRIPTION
**Pull request by Coconutcoo**
Fixes #45 

image

I *believe*, but can't confirm, that FS assigns every AccountName a discriminator, so I'm using that as a warning of the end of the AccountName.

If someone has a user from some other platform or something that doesn't have this, please let me know as I'll need to rethink.

Also, this PR has some tidying involved with the "get user from local storage" function, since it was conflicting with a non-exported function from useLocalStorage.

### CODE REVIEW

**Comment by Coconutcoo** [comment]
Please hang fire on a review - I know that safeUserParse doesn't decode the string right now - need to adjust the tests accordingly.

**Comment by danreeves** [comment]
Looking good so far

> I believe, but can't confirm, that FS assigns every AccountName a discriminator

I think it's a pretty good bet all accounts have it

File: src/utils.test.ts [hunk 1]
```diff
@@ -1,49 +1,47 @@
...
-  return Buffer.from(str, "ascii").toString("base64")
-}
-
-function fromB64(str: string): string {
-  return Buffer.from(str, "base64").toString()
-}
-
-function encode(obj: Record<string, unknown>): string {
-  return fromB64(toB64(JSON.stringify(obj)))
-}
```
Comment (danreeves) [inline src/utils.test.ts]: This part was to mimic the mangling fatshark are doing. e.g.

```js
JSON.parse(`{"AccountName":"【パンツ】Pantsu#1234"}`)
```
is fine but
```js
JSON.parse(fromB64(toB64(`{"AccountName":"【パンツ】Pantsu#1234"}`)))
```
errors

File: src/utils.test.ts [hunk 2]
```diff
@@ -1,49 +1,47 @@
...
-  `【パンツ】Pantsu`,
-  `🎃 ASH  П (◣_◢) П  🎃`,
-  `七曜曜`,
-  `Dunesca ʕ•͡ᴥ•ʔ`,
+  `【パンツ】Pantsu#1234`,
+  `🎃 ASH  П (◣_◢) П  🎃#1234`,
+  `七曜曜#1234`,
+  `Dunesca ʕ•͡ᴥ•ʔ#1234`,
+  `Deadsiesthe•̪̀●́#1234`,
+  `Deadsiesthe"*�#1234`
```
Comment (danreeves) [inline src/utils.test.ts]: Shouldn't need the duplucate if the unicode mangling is added back in

File: src/utils.test.ts [hunk 3]
```diff
@@ -1,49 +1,47 @@
...
+    let userString = `{\"AccessToken\": \"${TestAccessToken}\", \"AccountName\":\"${user}\"}`
+    let decoded = safeUserParse(userString)
     t.not(decoded, undefined)
-    t.is(decoded!.TestValue, TestValue)
+    t.is(decoded!.AccessToken, TestAccessToken)
+    t.not(decoded?.AccountName, undefined)
+    t.not(decoded?.AccountName, '')
+    if (decoded?.AccountName) {
+      t.is(/[\w-#]+/.test(decoded?.AccessToken), true)
+    }
```
Comment (danreeves) [inline src/utils.test.ts]: If you use the non-null assertion operator (https://www.typescriptlang.org/docs/handbook/2/everyday-types.html#non-null-assertion-operator-postfix-) you should be able to remove the `if`

```suggestion
    t.not(decoded!.AccountName, undefined)
    t.not(decoded!.AccountName, '')
    t.is(/[\w-#]+/.test(decoded!.AccessToken), true)
```

Not a fan of assertions but I'm happy with them in tests. If AccessToken was actually not on the object we would want the tests to fail anyway

File: src/utils.test.ts [hunk 4]
```diff
@@ -1,49 +1,47 @@
...
+      `{\"AccessToken\":\"${TestAccessToken}\", \"AccountName\":\"${user}\"}`,
+      `{\"AccountName\":\"${user}\", \"AccessToken\":\"${TestAccessToken}\"}`,
+      `{\"RefreshToken\":\"${TestAccessToken}\", \"AccountName\":\"${user}\", \"AccessToken\":\"${TestAccessToken}\"}`
+    ]
+    for (let str of strings) {
+      let decoded = safeUserParse(str)
+      t.not(decoded, undefined)
+      t.not(decoded?.AccountName, undefined)
+      t.not(decoded?.AccountName, '')
+      if (decoded?.AccountName) {
```
Comment (danreeves) [inline src/utils.test.ts]: Same as above

**Review by danreeves — CHANGES_REQUESTED** [review]
Looks good. just some comments on the tests

File: src/utils.test.ts (hunk 1 above)
Comment (Coconutcoo) [reply src/utils.test.ts]: Ah, you're right - I didn't realise you'd already basically worked out how they're mangling it, would've saved me some time!

Will re-add and give it a comment for future users

File: src/utils.test.ts (hunk 3 above)
Comment (Coconutcoo) [reply src/utils.test.ts]: TIL - I think I've seen that once or twice before perhaps, but nice to see a use in the wild for it

**Review by danreeves — APPROVED** [review]
Nice, thanks!

**Pull request merged**